File: integrations/codex/plugins/cognee/scripts/exit-watcher.py
#!/usr/bin/env python3
"""Run Cognee graph sync after the owning Codex process exits.

Codex CLI currently may not invoke plugin SessionEnd on normal shutdown.
SessionStart launches this watcher with the hook parent PID. The watcher
does nothing while Codex is alive; once that PID disappears, it starts the
normal detached graph sync worker and exits.
"""

import json
import os
import subprocess
import sys
import time
from pathlib import Path

_PLUGIN_DIR = Path.home() / ".cognee-plugin" / "codex"
_EXIT_WATCHERS_DIR = _PLUGIN_DIR / "exit-watchers"
_PIDFILE = _PLUGIN_DIR / "exit-watcher.pid"
_LOGFILE = _PLUGIN_DIR / "exit-watcher.log"
_SYNC_SCRIPT = Path(__file__).with_name("sync-session-to-graph.py")
_DETACHED_SYNC_ARG = "--detached-final"
_POLL_SECONDS = 2.0
_SYNC_START_DELAY = 2.0


def _log(event: str, **detail) -> None:
    try:
        _PLUGIN_DIR.mkdir(parents=True, exist_ok=True)
        line = {"ts": time.time(), "pid": os.getpid(), "event": event}
        if detail:
            line["detail"] = detail
        with _LOGFILE.open("a", encoding="utf-8") as fh:
            fh.write(json.dumps(line, default=str) + "\n")
    except Exception:
        pass


def _pid_alive(pid: int) -> bool:
    if pid <= 1:
        return False
    try:
        os.kill(pid, 0)
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True
    except Exception as exc:
        _log("pid_alive_check_failed", parent_pid=pid, error=str(exc)[:200])
        return False


def _owns_pidfile(pidfile: Path) -> bool:
    try:
        return int(pidfile.read_text(encoding="utf-8").strip()) == os.getpid()
    except Exception as exc:
        _log("pidfile_read_failed", pidfile=str(pidfile), error=str(exc)[:200])
        return False


def _spawn_sync(
    session_id: str,
    dataset: str,
    *,
    session_key: str = "",
    agent_session_name: str = "",
    api_key: str = "",
    service_url: str = "",
) -> None:
    try:
        env = os.environ.copy()
        env.setdefault("COGNEE_SYNC_START_DELAY", str(_SYNC_START_DELAY))
        env["COGNEE_UNREGISTER_ON_FINISH"] = "1"
        if session_id:
            env["COGNEE_SYNC_SESSION_ID"] = session_id
        if dataset:
            env["COGNEE_SYNC_DATASET"] = dataset
        if session_key:
            env["COGNEE_SESSION_KEY"] = session_key
        if agent_session_name:
            env["COGNEE_AGENT_SESSION_NAME"] = agent_session_name
        if api_key:
            env["COGNEE_API_KEY"] = api_key
        if service_url:
            env["COGNEE_BASE_URL"] = service_url
        subprocess.Popen(
            [sys.executable, str(_SYNC_SCRIPT), _DETACHED_SYNC_ARG],
            cwd=os.getcwd(),
            env=env,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        _log("exit_sync_deferred", session=session_id, dataset=dataset)
    except Exception as exc:
        _log("exit_sync_detach_failed", error=str(exc)[:300])


def main() -> None:
    if len(sys.argv) < 2:
        _log("fatal_missing_args")
        return
    try:
        bootstrap = json.loads(sys.argv[1])
    except Exception as exc:
        _log("fatal_bad_args", error=str(exc)[:200])
        return

    parent_pid = int(bootstrap.get("parent_pid") or 0)
    session_id = str(bootstrap.get("session_id") or "")
    dataset = str(bootstrap.get("dataset") or "agent_sessions")
    session_key = str(bootstrap.get("session_key") or "")
    agent_session_name = str(bootstrap.get("agent_session_name") or "")
    api_key = str(bootstrap.get("api_key") or "")
    service_url = str(bootstrap.get("base_url") or "")
    pidfile_raw = str(bootstrap.get("pidfile") or "").strip()
    pidfile = Path(pidfile_raw) if pidfile_raw else _PIDFILE
    if not parent_pid:
        _log("fatal_no_parent_pid", session=session_id, dataset=dataset)
        return

    try:
        _PLUGIN_DIR.mkdir(parents=True, exist_ok=True)
        _EXIT_WATCHERS_DIR.mkdir(parents=True, exist_ok=True)

        if pidfile.exists():
            try:
                existing = int(pidfile.read_text(encoding="utf-8").strip())
                if _pid_alive(existing):
                    _log(
                        "already_running_for_parent",
                        parent_pid=parent_pid,
                        session=session_id,
                        dataset=dataset,
                        pidfile=str(pidfile),
                        existing_pid=existing,
                    )
                    return
            except Exception:
                pass

        pidfile.write_text(str(os.getpid()), encoding="utf-8")
    except Exception as exc:
        _log("pidfile_write_failed", pidfile=str(pidfile), error=str(exc)[:200])
        return

    _log(
        "started",
        parent_pid=parent_pid,
        session=session_id,
        dataset=dataset,
        pidfile=str(pidfile),
    )
    while _owns_pidfile(pidfile) and _pid_alive(parent_pid):
        time.sleep(_POLL_SECONDS)

    if not _owns_pidfile(pidfile):
        _log("pidfile_replaced", parent_pid=parent_pid, pidfile=str(pidfile))
        return

    _log("parent_exited", parent_pid=parent_pid, session=session_id, dataset=dataset)
    _spawn_sync(
        session_id,
        dataset,
        session_key=session_key,
        agent_session_name=agent_session_name,
        api_key=api_key,
        service_url=service_url,
    )

    try:
        if _owns_pidfile(pidfile):
            pidfile.unlink()
    except Exception as exc:
        _log("pidfile_unlink_failed", pidfile=str(pidfile), error=str(exc)[:200])
    _log("exiting", parent_pid=parent_pid, pidfile=str(pidfile))


if __name__ == "__main__":
    main()

File: integrations/vellum-assistant/src/exit-watcher.ts
/**
 * Run Cognee graph sync after the owning host process exits.
 *
 * Ported from exit-watcher.py. A background process that:
 *   - Polls whether the parent PID is alive
 *   - When parent exits, spawns the detached final sync worker
 *   - Uses Bun.spawn with start_new_session equivalent
 *
 * The host (Vellum Assistant) may not invoke plugin SessionEnd on normal
 * shutdown. The init/session-start hook launches this watcher with the
 * parent PID. The watcher does nothing while the host is alive; once that
 * PID disappears, it starts the normal detached graph sync worker and exits.
 */

import { existsSync, readFileSync, writeFileSync, unlinkSync, mkdirSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { spawn } from "bun";

import { hookLog, pluginStateDir } from "./plugin-common.ts";

// ─── Constants ────────────────────────────────────────────────────────────────

const POLL_SECONDS = 2.0;
const SYNC_START_DELAY = 2.0;

function exitWatchersDir(): string {
  return join(pluginStateDir(), "exit-watchers");
}

function logPath(): string {
  return join(pluginStateDir(), "exit-watcher.log");
}

function defaultPidfilePath(): string {
  return join(pluginStateDir(), "exit-watcher.pid");
}

// ─── PID liveness ─────────────────────────────────────────────────────────────

function pidAlive(pid: number): boolean {
  if (pid <= 1) return false;
  try {
    process.kill(pid, 0);
    return true;
  } catch (err: any) {
    if (err?.code === "ESRCH") return false; // ProcessLookupError
    if (err?.code === "EPERM") return true; // PermissionError — alive but not ours
    hookLog("exit_watcher_pid_check_failed", { pid, error: String(err).slice(0, 200) });
    return false;
  }
}

function ownsPidfile(pidfile: string): boolean {
  try {
    return parseInt(readFileSync(pidfile, "utf-8").trim(), 10) === process.pid;
  } catch {
    return false;
  }
}

// ─── Detached sync spawn ──────────────────────────────────────────────────────

/**
 * Spawn the detached final sync worker (sync-session-to-graph --session-end).
 *
 * Uses Bun.spawn with a new session group so it survives the watcher exiting.
 */
export function spawnDetachedSync(params: {
  sessionId: string;
  dataset: string;
  sessionKey?: string;
  agentSessionName?: string;
  apiKey?: string;
  serviceUrl?: string;
}): void {
  const {
    sessionId,
    dataset,
    sessionKey = "",
    agentSessionName = "",
    apiKey = "",
    serviceUrl = "",
  } = params;

  try {
    const env: Record<string, string> = { ...process.env } as Record<string, string>;
    if (!env.COGNEE_SYNC_START_DELAY) {
      env.COGNEE_SYNC_START_DELAY = String(SYNC_START_DELAY);
    }
    env.COGNEE_UNREGISTER_ON_FINISH = "1";
    if (sessionId) env.COGNEE_SYNC_SESSION_ID = sessionId;
    if (dataset) env.COGNEE_SYNC_DATASET = dataset;
    if (sessionKey) env.COGNEE_SESSION_KEY = sessionKey;
    if (agentSessionName) env.COGNEE_AGENT_SESSION_NAME = agentSessionName;
    if (apiKey) env.COGNEE_API_KEY = apiKey;
    if (serviceUrl) env.COGNEE_BASE_URL = serviceUrl;

    // Resolve the sync-session-to-graph module path.
    // We re-exec this module file with the --detached-final flag, which
    // triggers the runSync(true) path.
    const modulePath = new URL(import.meta.url).pathname;
    const proc = spawn({
      cmd: ["bun", "run", modulePath, "--detached-final"],
      cwd: process.cwd(),
      stdin: "null",
      stdout: "null",
      stderr: "null",
      env,
      // start_new_session equivalent — detached so it survives the watcher.
      detached: true,
    });

    hookLog("exit_sync_deferred", { session: sessionId, dataset, pid: proc.pid });
  } catch (err) {
    hookLog("exit_sync_detach_failed", { error: String(err).slice(0, 300) });
  }
}

// ─── Watcher entry point ──────────────────────────────────────────────────────

export interface ExitWatcherBootstrap {
  parent_pid: number;
  session_id: string;
  dataset: string;
  session_key?: string;
  agent_session_name?: string;
  api_key?: string;
  base_url?: string;
  pidfile?: string;
}

/**
 * Run the exit watcher loop. Polls the parent PID until it exits,
 * then spawns the detached final sync worker.
 *
 * This is the background daemon body. It blocks until the parent exits.
 */
export async function runExitWatcher(bootstrap: ExitWatcherBootstrap): Promise<void> {
  const parentPid = bootstrap.parent_pid || 0;
  const sessionId = bootstrap.session_id || "";
  const dataset = bootstrap.dataset || "agent_sessions";
  const sessionKey = bootstrap.session_key || "";
  const agentSessionName = bootstrap.agent_session_name || "";
  const apiKey = bootstrap.api_key || "";
  const serviceUrl = bootstrap.base_url || "";
  const pidfileRaw = bootstrap.pidfile?.trim() || "";
  const pidfile = pidfileRaw || defaultPidfilePath();

  if (!parentPid) {
    hookLog("exit_watcher_fatal_no_parent_pid", { session: sessionId, dataset });
    return;
  }

  // Ensure directories exist.
  try {
    mkdirSync(pluginStateDir(), { recursive: true });
    mkdirSync(exitWatchersDir(), { recursive: true });
  } catch {
    // Best-effort.
  }

  // Check for an already-running watcher for this parent.
  try {
    if (existsSync(pidfile)) {
      const existing = parseInt(readFileSync(pidfile, "utf-8").trim(), 10);
      if (pidAlive(existing)) {
        hookLog("exit_watcher_already_running", {
          parent_pid: parentPid,
          session: sessionId,
          existing_pid: existing,
        });
        return;
      }
    }
    writeFileSync(pidfile, String(process.pid), "utf-8");
  } catch (err) {
    hookLog("exit_watcher_pidfile_write_failed", { pidfile, error: String(err).slice(0, 200) });
    return;
  }

  hookLog("exit_watcher_started", {
    parent_pid: parentPid,
    session: sessionId,
    dataset,
    pidfile,
  });

  // Poll until the parent exits or we lose pidfile ownership.
  while (ownsPidfile(pidfile) && pidAlive(parentPid)) {
    await sleep(POLL_SECONDS * 1000);
  }

  if (!ownsPidfile(pidfile)) {
    hookLog("exit_watcher_pidfile_replaced", { parent_pid: parentPid, pidfile });
    return;
  }

  hookLog("exit_watcher_parent_exited", { parent_pid: parentPid, session: sessionId, dataset });

  // Spawn the detached final sync worker.
  spawnDetachedSync({
    sessionId,
    dataset,
    sessionKey,
    agentSessionName,
    apiKey,
    serviceUrl,
  });

  // Clean up the pidfile.
  try {
    if (ownsPidfile(pidfile)) {
      unlinkSync(pidfile);
    }
  } catch (err) {
    hookLog("exit_watcher_pidfile_unlink_failed", { pidfile, error: String(err).slice(0, 200) });
  }

  hookLog("exit_watcher_exiting", { parent_pid: parentPid, pidfile });
}

/**
 * Launch the exit watcher as a detached background process.
 *
 * Called from session-start. The watcher re-execs this module file with
 * the bootstrap JSON as an argument.
 *
 * @param bootstrap The watcher configuration.
 * @returns The spawned process PID, or 0 on failure.
 */
export function launchExitWatcher(bootstrap: ExitWatcherBootstrap): number {
  const pidfileRaw = bootstrap.pidfile?.trim() || "";
  const pidfile = pidfileRaw || join(exitWatchersDir(), `${bootstrap.parent_pid}.pid`);

  // Prune stale watcher pidfiles.
  try {
    if (existsSync(exitWatchersDir())) {
      for (const name of readdirSync(exitWatchersDir()) as string[]) {
        if (!name.endsWith(".pid")) continue;
        const p = join(exitWatchersDir(), name);
        try {
          const pid = parseInt(readFileSync(p, "utf-8").trim(), 10);
          if (!pidAlive(pid)) {
            unlinkSync(p);
          }
        } catch {
          continue;
        }
      }
    }
  } catch (err) {
    hookLog("exit_watcher_prune_failed", { error: String(err).slice(0, 200) });
  }

  // Check if a watcher is already running for this parent.
  try {
    if (existsSync(pidfile)) {
      const existing = parseInt(readFileSync(pidfile, "utf-8").trim(), 10);
      if (pidAlive(existing)) {
        hookLog("exit_watcher_already_running", {
          parent_pid: bootstrap.parent_pid,
          pidfile,
        });
        return 0;
      }
    }
  } catch {
    // Fall through.
  }

  // Update the bootstrap with the resolved pidfile path.
  const fullBootstrap = { ...bootstrap, pidfile };

  try {
    const env: Record<string, string> = { ...process.env } as Record<string, string>;
    if (bootstrap.session_key) {
      env.COGNEE_SESSION_KEY = bootstrap.session_key;
    }

    const modulePath = new URL(import.meta.url).pathname;
    const proc = spawn({
      cmd: ["bun", "run", modulePath, JSON.stringify(fullBootstrap)],
      stdin: "null",
      stdout: logPath(),
      stderr: logPath(),
      env,
      detached: true,
    });

    hookLog("exit_watcher_started", {
      parent_pid: bootstrap.parent_pid,
      session: bootstrap.session_id,
      dataset: bootstrap.dataset,
      pidfile,
      pid: proc.pid,
    });

    return proc.pid ?? 0;
  } catch (err) {
    hookLog("exit_watcher_launch_failed", { error: String(err).slice(0, 300) });
    return 0;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ─── CLI entry point ──────────────────────────────────────────────────────────

/**
 * When this module is run directly (e.g. `bun run exit-watcher.ts <bootstrap>`),
 * parse the bootstrap JSON arg and run the watcher loop.
 *
 * Also handles the `--detached-final` arg from spawnDetachedSync, which
 * delegates to the sync module.
 */
async function main(): Promise<void> {
  const args = process.argv.slice(2);

  // --detached-final: delegate to sync-session-to-graph's runSync(true).
  if (args.includes("--detached-final")) {
    try {
      const { runSync } = await import("./sync-session-to-graph.ts");
      await runSync(true);
    } catch (err) {
      hookLog("exit_watcher_detached_sync_error", { error: String(err).slice(0, 300) });
    }
    return;
  }

  // Normal watcher mode: bootstrap JSON is the first arg.
  if (args.length < 1) {
    hookLog("exit_watcher_fatal_missing_args");
    return;
  }

  try {
    const bootstrap = JSON.parse(args[0]) as ExitWatcherBootstrap;
    await runExitWatcher(bootstrap);
  } catch (err) {
    hookLog("exit_watcher_fatal_bad_args", { error: String(err).slice(0, 200) });
  }
}

// Run main when this file is the entry point.
if (import.meta.main) {
  main();
}

export default runExitWatcher;
